File: ios/BetterW4/W4ClassId.swift
import Foundation

/// W4 class identifiers as printed on the timetable brick and in My classes.
///
/// Live capture (nc26jban, Aug 2026): `1DA13HMTAA` = 1st year, block D, room A 1.3,
/// Higher, Mathematics Analysis and Approaches. The brick shows only this code;
/// the real name is in `div.period[title]` as `Class: <b>…</b>`.
struct W4ClassId: Equatable, Sendable {
    let raw: String
    let year: Int
    let block: String
    let roomCode: String
    let level: Character
    let subjectCode: String

    var levelLabel: String {
        switch level {
        case "H": return "HL"
        case "S": return "SL"
        case "C": return "C"
        case "X": return "X"
        default: return String(level)
        }
    }

    private static let pattern = #"^(\d)([A-Za-z])([A-Za-z]{1,3}\d{0,2})([HSCXhscx])([A-Za-z]{3,5})$"#

    static func looksLike(_ raw: String) -> Bool { parse(raw) != nil }

    static func parse(_ raw: String) -> W4ClassId? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let regex = try? NSRegularExpression(pattern: pattern),
              let match = regex.firstMatch(
                in: trimmed,
                range: NSRange(trimmed.startIndex..<trimmed.endIndex, in: trimmed)
              ),
              match.numberOfRanges == 6
        else {
            return nil
        }

        func group(_ index: Int) -> String? {
            guard let range = Range(match.range(at: index), in: trimmed) else { return nil }
            return String(trimmed[range])
        }

        guard let yearText = group(1), let year = Int(yearText),
              let block = group(2),
              let room = group(3),
              let levelText = group(4), let level = levelText.uppercased().first,
              let subject = group(5)
        else {
            return nil
        }

        return W4ClassId(
            raw: trimmed,
            year: year,
            block: block.uppercased(),
            roomCode: room.uppercased(),
            level: level,
            subjectCode: subject.uppercased()
        )
    }
}

File: android/app/src/main/java/dk/betterw4/android/feature/schedule/W4ClassId.kt
package dk.betterw4.android.feature.schedule

/**
 * W4 class identifiers as printed on the timetable brick and in My classes.
 *
 * Live capture (nc26jban, Aug 2026): `1DA13HMTAA`, `1EA16CECOX`, `1YA25SLALI`,
 * `1ZAUDXCORE`. The compact code is:
 *
 *     {year}{block}{room}{level}{subject}
 *
 *     1 DA13 H MTAA  →  1st year, block D, room A 1.3, Higher, Mathematics AA
 *     1 EA16 C ECOX  →  1st year, block E, room A 1.6, C level, Economics
 *     1 ZAUD X CORE  →  1st year, block Z, Auditorium, X level, Core meetings
 *
 * The brick itself shows only this code. The real subject name lives in
 * `div.period[title]` as `Class: <b>Economics</b>` and in My classes as the
 * `<dt>` heading.
 *
 * Advisor groups use a first name (`Dona`, `Josef`) instead of this shape.
 */
data class W4ClassId(
    val raw: String,
    val year: Int,
    val block: String,
    val roomCode: String,
    val level: Char,
    val subjectCode: String,
) {
    val levelLabel: String
        get() = when (level.uppercaseChar()) {
            'H' -> "HL"
            'S' -> "SL"
            'C' -> "C"
            'X' -> "X"
            else -> level.toString()
        }

    companion object {
        private val PATTERN = Regex(
            """^(\d)([A-Za-z])([A-Za-z]{1,3}\d{0,2})([HSCXhscx])([A-Za-z]{3,5})$""",
        )

        fun parse(raw: String): W4ClassId? {
            val trimmed = raw.trim()
            val match = PATTERN.matchEntire(trimmed) ?: return null
            return W4ClassId(
                raw = trimmed,
                year = match.groupValues[1].toInt(),
                block = match.groupValues[2].uppercase(),
                roomCode = match.groupValues[3].uppercase(),
                level = match.groupValues[4].uppercase()[0],
                subjectCode = match.groupValues[5].uppercase(),
            )
        }

        fun looksLike(raw: String): Boolean = parse(raw) != null
    }
}
